KFSPTS-7912: Initial commit for implementing SAE-to-Collector processing.

cah292 opened the pull request:
This is an incomplete implementation of the SAE-to-Collector processing, which I'm trying to move into develop so that I can split up some of the remaining work. The code is now at the point where it can take the SAE POJOs and convert them into GL Collector BOs; however, that portion has not been tied into the needed SAE services for execution in a runtime environment, and more changes to it (and its related micro-tests) will be coming soon once the SAE specs are revised and clarified. In addition, the BO-to-file portion has not been included in this PR since it's not in a state to be merged yet.

This PR references some new parameters that I'm planning to use, but I haven't created the SQL for them yet since the new code is not currently used at runtime. The SQL will coincide with a future PR, unless you think it should coincide with this PR.

jhulslander commented on src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurStandardAccountingExtractCreateCollectorFileServiceImpl.java:
```diff
@@ -0,0 +1,61 @@
+        return true;
+    }
+
+    protected String getConcurParameterValueAsString(String parameterName) {
```
I've used this function a couple of times, I think Daniela has as well.  Would it be a good idea to move this into a shared service or utility so it isn't duplicated?

cah292 replied:
Yes, moving to a new service or utility sounds like a good idea. I will look into adding that.

nkimble replied:
KFSPTS-7997 has already been created as a future user story for that type of consolidation.
I would suggest marking that method with a comment for now and we do this shared service change as part of that work

cah292 replied:
Thanks for checking, Nancy. I'll add comments accordingly so that the future consolidation work will clean this up.

jhulslander replied:
Thanks Nancy, good plan

jhulslander commented on src/main/java/edu/cornell/kfs/gl/CuGeneralLedgerConstants.java:
```diff
@@ -5,6 +5,7 @@
 public class CuGeneralLedgerConstants extends GeneralLedgerConstants {
 
 	public static final String ANNUAL_CLOSING_CHARTS_PARAM = "ANNUAL_CLOSING_CHARTS";
+	public static final String COLLECTOR_HEADER_RECORD_TYPE = "HD";
```
It feels like GL Constants is a better place. I don't have a terribly strong opinion on this matter

cah292 replied:
This is the same value that the existing GL Collector parsing looks for when parsing. The base KFS code doesn't appear to have it defined in a variable (or at least in a publicly-available variable), which is why I added it to our GL constants.

db483 replied:
So this constant is currently not used in any of the code on this PR?

cah292 replied:
I'm using this constant when initializing the CollectorBatch object in the builder class. I'm also using it for the unit test code that builds a CollectorBatch from a fixture.

db483 replied:
Since this a constant for a hard coded value used by the Collector related code in the GL module as well I think it makes sense to keep it in this file.

Initially I thought this was a new value used only by the Concur related code hence the suggestion to move it in the Concur module.

jhulslander commented on src/test/java/edu/cornell/kfs/concur/ConcurTestConstants.java:
```diff
@@ -0,0 +1,49 @@
+package edu.cornell.kfs.concur;
+
+public class ConcurTestConstants {
```
I only checked a couple, but if these constants are only used in one fixture, do we need a separate constants est class?

cah292 replied:
Some of these constants are used by at least two different fixtures (and in multiple spots), so I defined them in this spot to cut down on duplication. Also, since my fixtures are enum-based, I don't think I can define them above the enum constants if I wanted to keep them in the same file. I also wanted to keep things consolidated here in case future SAE-to-Collector micro-test classes are added.

jhulslander replied:
OK, thanks

jhulslander commented on src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurStandardAccountingExtractCollectorBatchBuilder.java:
```diff
@@ -0,0 +1,304 @@
+    protected CollectorBatch collectorBatch;
+    protected Map<String,OriginEntryFull> originEntries;
+    protected Function<Date,Integer> fiscalYearFinder;
+    protected BiFunction<Date,String,String> dateFormatter;
```
I haven't seen this before, would this be a good item for a book club discussion?  It's a cool method

cah292 replied:
It's just some usage of Java 8's new functional interfaces, but if you guys want to discuss it at a bookclub meeting prior to us starting any Java-8-related books, then I could discuss this there.

jhulslander commented on src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurStandardAccountingExtractCreateCollectorFileServiceImpl.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    public boolean buildCollectorFile(ConcurStandardAccountingExtractFile saeFileContents) {
+        ConcurStandardAccountingExtractCollectorBatchBuilder builder = new ConcurStandardAccountingExtractCollectorBatchBuilder(
+                universityDateService::getFiscalYear, dateTimeService::toString,
```
would it be better to define these lamba functions as a variables to help with understanding what is going on?

cah292 replied:
I'll see if I can change some of this over to passing in actual services instead of method references, which should help clean this up. I may not need this much abstraction anymore, now that some other Concur PRs have been merged.

cah292 commented on src/main/java/edu/cornell/kfs/concur/ConcurParameterConstants.java:
```diff
@@ -10,5 +10,14 @@
     public static final String CONCUR_PENDING_CLIENT_OBJECT_CODE_OVERRIDE = "CONCUR_PENDING_CLIENT_OBJECT_CODE_OVERRIDE";
-
+    public static final String CONCUR_SAE_COLLECTOR_DOCUMENT_TYPE = "CONCUR_SAE_COLLECTOR_DOCUMENT_TYPE";
+    public static final String CONCUR_SAE_COLLECTOR_SYSTEM_ORIGINATION_CODE = "CONCUR_SAE_COLLECTOR_SYSTEM_ORIGINATION_CODE";
```
As per the specs, the SAE-to-PDP origination code should be "Z6", but the SAE-to-Collector origination code should be "CN" instead. Thus, we do need a separate parameter for this.

db483 replied:
Thank you for clarifying this Chad.

cah292 commented on src/main/java/edu/cornell/kfs/concur/ConcurParameterConstants.java:
```diff
@@ -10,5 +10,14 @@
+    public static final String CONCUR_SAE_COLLECTOR_DEPARTMENT_NAME = "CONCUR_SAE_COLLECTOR_DEPARTMENT_NAME";
+    public static final String CONCUR_SAE_COLLECTOR_CAMPUS_CODE = "CONCUR_SAE_COLLECTOR_CAMPUS_CODE";
+    public static final String CONCUR_SAE_COLLECTOR_CAMPUS_ADDRESS = "CONCUR_SAE_COLLECTOR_CAMPUS_ADDRESS";
+    public static final String CONCUR_SAE_COLLECTOR_NOTIFICATION_CONTACT_EMAIL = "CONCUR_SAE_COLLECTOR_NOTIFICATION_CONTACT_EMAIL";
```
I was planning to use parameters for these values, since the specs already mention which values should be used. I think only a few of these are in the customer profile, and the customer profile is in PDP so I'm not sure if it would be that useful for this GL processing.

cah292 commented on src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurStandardAccountingExtractCollectorBatchBuilder.java:
```diff
@@ -0,0 +1,304 @@
+        
+        return new StringBuilder(INITIAL_BUILDER_SIZE)
+                .append(makeEmptyIfBlank(saeLine.getChartOfAccountsCode()))
+                .append('|').append(makeEmptyIfBlank(saeLine.getAccountNumber()))
```
Ok, I'll move this to its own constant.

cah292 commented on src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurStandardAccountingExtractCollectorBatchBuilder.java:
```diff
@@ -0,0 +1,304 @@
+        
+        return new StringBuilder(INITIAL_BUILDER_SIZE)
+                .append(StringUtils.left(saeLine.getEmployeeLastName(), NAME_LENGTH_FOR_DESCRIPTION))
+                .append(',').append(StringUtils.left(saeLine.getEmployeeFirstName(), NAME_LENGTH_FOR_DESCRIPTION))
```
Ok, I'll move this to its own constant.

cah292 commented:
I've made some more changes based on feedback, plus a few other tweaks. I've also created an associated nonprod-sql PR; please make sure both PRs are good to go before merging them.

I know that more changes are needed based on today's meeting with Sherry, but I will be addressing them in a follow-up user story.

This is now ready to be reviewed again.

db483 commented on src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurStandardAccountingExtractCollectorBatchBuilder.java:
```diff
@@ -0,0 +1,310 @@
+            case ConcurConstants.PAYMENT_CODE_UNIVERSITY_BILLED_OR_PAID :
+                return true;
+            case ConcurConstants.PAYMENT_CODE_PRE_PAID_OR_OTHER :
+                LOG.warn("Found a transaction with a Pre-Paid/Other payment code; it will not be processed. Item key: " + itemKey);
```
should these logs be debug or info? I am trying to understand why they are warnings. Are these types of entries not expected in the SAE file?

cah292 replied:
As per the specs, we will eventually have to add some more sophisticated error reporting for the Concur file extracts. In the meantime, I've added some logging statements like this, in case we need to find the related information prior to the reporting being implemented. I put them at WARN level to help them stand out as part of that effort; if you think INFO would be better, then I can make that change.

cah292 commented:
I was planning on having this code merged in as-is (unless anyone notices something that absolutely needs to be addressed first). I will be implementing the spec-related changes and the BO-to-flat-file processing in a subsequent user story.

cah292 commented:
I think I've addressed the comments and concerns for this. If there's still something more that should be done prior to merging this PR or its related PR, please let me know.

cah292 commented on src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurStandardAccountingExtractCollectorBatchBuilder.java:
```diff
@@ -0,0 +1,310 @@
+    protected int batchSequenceNumber;
+    protected int nextFakeObjectCode;
+
+    // NOTE: The last argument needs to be updated to use an actual service once one has been implemented.
```
Ok, I'll remove the comment.

cah292 commented on src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurStandardAccountingExtractCreateCollectorFileServiceImpl.java:
```diff
@@ -0,0 +1,61 @@
+        return true;
+    }
+
+    // This method needs to be replaced with an actual separate service once one has been implemented.
```
Ok, I'll remove the comment.

cah292 commented:
I've made some more tweaks based on the latest feedback. This is ready for review again.

cah292 commented:
One of the recently-merged PRs has just caused some merge conflicts on this one. I'll need to rebase accordingly before this can be merged.

cah292 commented:
I've rebased this branch, so it is now in a mergable state. This is now ready for review again.

cah292 commented:
retest this please

cah292 commented:
Does everyone approve of these changes? Can they be merged yet?

jhulslander commented:
Yes, changes look good to me

nkimble commented:
Changes are ok.

cah292 commented:
Thanks!
